output/cloud/output.go

```diff
@@ -283,9 +291,37 @@ func (out *Output) Description() string {
 	return fmt.Sprintf("cloud (%s)", cloudapi.URLForResults(out.referenceID, out.config))
 }
 
-// SetRunStatus receives the latest run status from the Engine.
-func (out *Output) SetRunStatus(status lib.RunStatus) {
-	out.runStatus = status
+// getRunStatus determines the run status of the test based on the error.
+func (out *Output) getRunStatus(testErr error) cloudapi.RunStatus {
+	if testErr == nil {
+		return cloudapi.RunStatusFinished
+	}
+
+	var err errext.HasAbortReason
+	if errors.As(testErr, &err) {
+		abortReason := err.AbortReason()
+		switch abortReason {
+		case errext.AbortedByUser:
+			return cloudapi.RunStatusAbortedUser
+		case errext.AbortedByThreshold:
+			return cloudapi.RunStatusAbortedThreshold
+		case errext.AbortedByScriptError:
+			return cloudapi.RunStatusAbortedScriptError
+		case errext.AbortedByScriptAbort:
+			return cloudapi.RunStatusAbortedUser // TODO: have a better value than this?
+		case errext.AbortedByTimeout:
+			return cloudapi.RunStatusAbortedLimit
+		case errext.AbortedByThresholdsAfterTestEnd:
+			// The test finished normally, it wasn't prematurely aborted. Such
+			// failures are tracked by the restult_status, not the run_status
+			// (so called "tainted" in some places of the API here).
+			return cloudapi.RunStatusFinished
```

Comment:
I improved the comments in https://github.com/grafana/k6/pull/2810/commits/be3aa5f89c854dece753287783eb07c7d46af084, which hopefully clears up some things.

The TLDR versions is that we have a mismatch between the k6 OSS and cloud. k6 OSS needs to have an error here, so it can exit with a non-0 exit code, but the k6 cloud tracks that error differently and needs the normal `RunStatusFinished` here :disappointed: Though I am also not super with the `AbortedByThresholdsAfterTestEnd` name, so if anyone has any better suggestions, please comment here: https://github.com/grafana/k6/blob/be3aa5f89c854dece753287783eb07c7d46af084/errext/abort_reason.go#L13

Reply:
Thanks, but the comment was the least of my concerns here :sweat_smile: While a long explanation of this mismatch helps, I'd prefer if we could avoid it.

My main objection is that if thresholds fail after the test ends, the test wasn't _aborted_. So it's not so much about the naming of `AbortedByThresholdsAfterTestEnd`, but that it's wrong and confusing, and that it shouldn't be an `AbortReason` at all. The only reason we use it, is so we can easily catch it here, but we can avoid this if we detect this scenario in a different way.

Maybe use a different error altogether? I'm not sure what the best approach would be, but the current situation is confusing even with the long explanation.

Reply:
> but we can avoid this if we detect this scenario in a different way.

The problem is that this will require a very, very substantial refactoring, just for this one single minor mismatch... :disappointed: And we'd essentially end up needing two different mechanisms for reporting the error to the output, or something like that :thinking: 

But overall it is a very fair point :thinking: `AbortReason` is not that great of a name for other reasons, now that I think about it some more, especially if we want to use it for https://github.com/grafana/k6/pull/2810#discussion_r1083770450 :thinking:  What do you think about `TestRunError` as the type name and these as the values:
```go

// TestRunError is used to attach extra information for the cause of a test run failure or interruption.
type TestRunError uint8

const (
	TestRunAbortedByUser TestRunError = iota + 1
	TestRunAbortedByBreachedThresholds
	TestRunThresholdsWereBreached // at the end of the test
	TestRunAbortedByScriptError
	TestRunAbortedByScriptAbort
	TestRunAbortedByTimeout
)

// IsTestRunError is a wrapper around an error with an attached TestRunError.
type IsTestRunError interface {
	error
	TestRunError() TestRunError
}
```

Reply:
:thinking: Yeah, `TestRunError` is a better name than `AbortReason`, but it seems like we're just prefixing `TestRun` to make it more generic so we can fit `TestRunThresholdsWereBreached` :smile:

The `cloud` output already knows if thresholds were breached with [this "tainted" check](https://github.com/grafana/k6/blob/be3aa5f89c854dece753287783eb07c7d46af084/output/cloud/output.go#L675-L685). It could arguably even avoid this check, since the `Engine` (or whatever we end up moving this to in this series of PRs) already knows this from [`engine.IsTainted()`](https://github.com/grafana/k6/blob/be3aa5f89c854dece753287783eb07c7d46af084/cmd/run.go#L273), and the output could just query it. In the meantime, could it infer the `RunStatus` from this?

I feel like we're abusing errors as a way to pass information to outputs, and making global changes that are only useful for the `cloud` output. This could be avoided if we had an API for outputs to query this and other information instead, and it's related to the discussion below about `StopWithTestError()`.

If you think this would be a much larger refactor, then let's use `TestRunError` for now, but we should discuss and improve this soon-ish.

Reply:
Yeah, it will be quite a big refactor, and it will be difficult to make an API that doesn't have timing issues. This API will only work once the test has finished and the thresholds have been crunched. But the output will only know that this has happened once its `Stop` method has been called. So my thinking was - why not pass that information to taht method anyway? :sweat_smile: 

Even my original idea from https://github.com/grafana/k6/issues/2430, to have outputs know that they need to stop when their `Samples` channel is closed, is not very workable... :disappointed: After we finish the test and run out of samples, we first need to crunch the thresholds and only then stop the rest of the outputs... So yeah, this is complicated and I'd very much like to leave solving it for after we have gotten rid of the `Engine`, since that will simplify things tremendously